docs/contributors/code/coding-guidelines.md

```diff
@@ -136,10 +136,278 @@ export { __unstableDoTerribleAwfulAction } from './api';
+		// ...
+	}
+```
+
+In addition, the existing public actions can dispatch private actions:
+
+```js
+export function toggleFeature( scope, featureName ) {
+    return function ( { dispatch } ) {
+		dispatch({ type: 'BEFORE_SAVE' })
+		// ...
+    };
+}
+```
+
+#### `lock()` and `unlock()`
+
+Use the `lock()` and `unlock()` API from `@wordpress/experiments` to
+privately export almost anything. See the examples below and learn more
+in the [`@wordpress/experiments` package README.md](https://github.com/WordPress/gutenberg/blob/HEAD/packages/experiments/README.md).
+
+##### Experimental selectors and actions
+
+Attach private selectors and actions to a public store:
+
+```js
+// In @wordpress/package1/store.js:
+import { __experimentalHasContentRoleAttribute } from './selectors';
+import { lock } from './experiments';
+
+export const store = registerStore(/* ... */);
+/* Attach a private selector to the exported store: */
+lock(store, {
+	__experimentalHasContentRoleAttribute
+});
+
+// In @wordpress/package2/MyComponent.js:
+import { store } from '@wordpress/package1';
+import { useSelect } from '@wordpress/data';
+import { unlock } from './experiments';
+
+function MyComponent() {
+    const hasRole = useSelect( ( select ) => (
+		// Use the private selector:
+        unlock( select( store ) ).__experimentalHasContentRoleAttribute()
+    ) );
+
+    // ...
+}
+```
+
+##### Experimental functions, classes, and variables
+
+```js
+// In @wordpress/package1/index.js:
+import { lock } from './experiments';
+
+export const experiments = {};
+/* Attach private data to the exported object */
+lock(store, {
+	__experimentalCallback: function() {},
+	__experimentalReactComponent: function MyComponent() { return <div/>; },
+	__experimentalClass: class Experiment{},
+	__experimentalVariable: 5,
+});
+
+
+// In @wordpress/package2/index.js:
+import { experiments } from '@wordpress/package1';
+import { unlock } from './experiments';
+
+const {
+	__experimentalCallback,
+	__experimentalReactComponent,
+	__experimentalClass,
+	__experimentalVariable
+} = unlock( experiments );
+```
+
+#### Experimental function arguments
+
+To add an experimental argument to a stable function you'll need
+to prepare a stable and an experimental version of that function.
+Then, export the stable function and `lock()` the unstable function
+inside it:
+
+```js
+// In @wordpress/package1/index.js:
+import { lock } from './experiments';
+
+export function log(data) {
+	__experimentalLog(data, 'log');
+}
+function __experimentalLog(data, __experimentalMethod) {
+	if(__experimentalMethod === 'log') {
+		console.log(data)
+	} else if(__experimentalMethod === 'table') {
+		console.table(data)
+	}
+}
+
+// In @wordpress/package2/index.js:
+import { log } from '@wordpress/package1';
+import { unlock } from './experiments';
+
+const __experimentalLog = unlock(log);
+__experimentalLog(data, 'table');
+// Or:
+unlock(log)('table');
+```
+
+#### Experimental React Component properties
```

Comment:
I also think this is misleading, as adding properties often (if not always) need to impact the internals of components, so you do a condition to render two separate trees unless you duplicate a lot of code.  Same for the functions.

Reply:
@youknowriad Whenever you have an experimental property, you also need a well-defined stable behavior. The pattern of "a full-featured private component" and "limited public facade that disables the experimental behavior" captures that dynamics. Do you have any counter-examples where that wouldn't work?

Reply:
I'd say most cases where we do experimental props don't work here (unless I'm missing something). You just add a prop and you want to change something in the rendered elements, example:

```js
const MyExistingButton = ( { title } ) => {
  // Do some complex stuff that I don't want to duplicate.
  
  return (
       <button>
           <Icon src={some icon} /> { title }  
      </button>
    );
}

```

now introduce the experimental prop

```
const MyExistingButton = ( { title, __experimentalShowIcon = true } ) => {
  // Do some complex stuff that I don't want to duplicate.
  
  return (
       <button>
           { __experimentalShowIcon  && <Icon src={some icon} /> } { title }  
      </button>
    );
}
```

How do I introduce "__experimentalShowIcon" without duplicating the whole component?

Reply:
@youknowriad Here's what I'd do:

```js
export const MyExistingButton = ( { title } ) => 
    <ExperimentalMyExistingButton title={ title } __experimentalShowIcon={ false } />

const ExperimentalMyExistingButton = ( { title, __experimentalShowIcon = true } ) => {
  // Do some complex stuff that I don't want to duplicate.
  
  return (
       <button>
           { __experimentalShowIcon  && <Icon src={some icon} /> } { title }  
      </button>
    );
}

lock(MyExistingButton, ExperimentalMyExistingButton);
```

Actually, if `__experimentalShowIcon` defaulted to false, we could do just:

```js
export const MyExistingButton = ( { title } ) => 
    <ExperimentalMyExistingButton title={ title } />
```

The public component does not accept any experimental props. The private component does accepts experimental props. The public component calls the private one such that the experimental features are disabled.

Reply:
Makes sense, I didn't think about inverting the responsibilities.

Reply:
It still feels weird though to have to define a second component to be able to add an argument to an existing one.